File: compiler/src/transform/expand-generics.ts
/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import {
  Interface,
  Model,
  TypeAlias,
  TypeDefinition,
  TypeName,
  ValueOf,
  Request,
  Response,
  Body,
  InstanceOf, Inherits, Property, ResponseException
} from '../model/metamodel'
import { readFile, writeFile } from 'fs/promises'
import stringify from 'safe-stable-stringify'
import { sortTypeDefinitions } from '../model/utils'
import { argv } from 'zx'
import { join } from 'path'

export class ExpansionConfig {
  unwrappedTypes?: TypeName[] | string[]
  inlinedTypes?: TypeName[] | string[]
}

/**
 * Expand all generics by creating new concrete types for every instantiation of a generic type.
 *
 * The resulting model has no generics anymore. Top-level generic parameters (e.g. SearchRequest's TDocument) are
 * replaced by user_defined_data.
 *
 * @param inputModel the input model
 * @param unwrappedTypes types that should not be expanded but unwrapped as their generic parameter.
 * @return a new model with generics expanded
 */
export function expandGenerics (inputModel: Model, config?: ExpansionConfig): Model {
  const typesToUnwrap = new Set<string>()
  const typesToInline: Set<string> = new Set<string>()

  for (const name of config?.unwrappedTypes ?? []) {
    if (typeof name === 'string') {
      typesToUnwrap.add(name)
    } else {
      typesToUnwrap.add(nameKey(name))
    }
  }

  for (const name of config?.inlinedTypes ?? []) {
    if (typeof name === 'string') {
      typesToInline.add(name)
    } else {
      typesToInline.add(nameKey(name))
    }
  }

  const typesSeen = new Set<string>()

  const types = new Array<TypeDefinition>()

  const inputTypeByName = new Map<string, TypeDefinition>()

  for (const type of inputModel.types) {
    inputTypeByName.set(nameKey(type), type)
  }

  /**
   * If a type has not already been seen, create it.
   * Note: this also set the `name` property of the result of `build()` (you don't have to care about setting it when
   * cloning a type definition).
   */
  function addIfNotSeen (name: TypeName, build: () => TypeDefinition): TypeName {
    const key = nameKey(name)
    if (!typesSeen.has(key)) {
      typesSeen.add(key)
      const type = build()
      type.name = name
      types.push(type)
    }

    return name
  }

  /**
   * Get a type by name in the input model, and fails if it doesn't exist
   */
  function getType (name: TypeName): TypeDefinition {
    const result = inputTypeByName.get(nameKey(name))
    if (result === undefined) {
      throw Error(`Type ${nameKey(name)} does not exist.`)
    } else {
      return result
    }
  }

  /**
   * Expand a root type (request or response) by replacing its generic parameters with user defined data
   */
  function expandRootType (name: TypeName | null): void {
    if (name == null) {
      return
    }

    const type = getType(name)
    if (type.kind !== 'request' && type.kind !== 'response') {
      throw Error(`${nameKey(name)} should be a request or a response`)
    }

    const userDefined: ValueOf = {
      kind: 'user_defined_value'
    }

    const typeParams = (type.generics ?? []).map(x => userDefined)
    expandType(type.name, typeParams)
  }

  /**
   * Expand a type, given concrete values for its generic parameters
   */
  function expandType (name: TypeName, params: ValueOf[]): TypeName {
    if (name.namespace === '_builtins') {
      return name
    }

    const type = getType(name)

    switch (type.kind) {
      case 'enum':
        // Nothing to do
        return addIfNotSeen(type.name, () => type)

      case 'type_alias':
        return expandTypeAlias(type, params)

      case 'request':
        return expandRequest(type, params)

      case 'response':
        return expandResponse(type, params)

      case 'interface':
        return expandInterface(type, params)
    }
  }

  /**
   * Add dangling types like CommonQueryParameters & BaseEsqlVersion to the generic less schema.
   * @param type the type definition
   */
  function addDanglingTypeIfNotSeen (type: TypeDefinition): void {
    switch (type.kind) {
      case 'type_alias':
        if (type.generics !== undefined && type.generics.length > 0) {
          return
        }
        break
      case 'interface':
        if (type.generics !== undefined && type.generics.length > 0) {
          return
        }
        break
    }
    addIfNotSeen(type.name, () => type)
  }

  /**
   * Expand an interface definition.
   *
   * @param type the type definition
   * @param params concrete values for the type's generic parameters, if any
   */
  function expandInterface (type: Interface, params: ValueOf[]): TypeName {
    return addIfNotSeen(expandedName(type, params), () => {
      const result = { ...type }

      const mappings = genericParamMapping(type.generics, params)

      result.inherits = expandInherits(result.inherits, mappings)

      // We add to the schema the non generics behaviors
      // CommonQueryParameters
      // CommonCatQueryParameters
      if (result.behaviors != null) {
        result.behaviors.forEach(b => {
          if (b.generics == null) {
            const type = getType(b.type)
            addIfNotSeen(b.type, () => type)
          }
        })
      }

      if (result.behaviors != null) {
        // We keep the generic parameters, but expand their value
        result.behaviors = result.behaviors.map(b => {
          return {
            type: b.type,
            generics: (b.generics ?? []).map(g => expandValueOf(g, mappings))
          }
        })
      }

      result.properties = expandProperties(result.properties, mappings)
      result.generics = undefined
      return result
    })
  }

  /**
   * Expand an inherits clause.
   *
   * @param inherits the inherits clause
   * @param mappings mapping of generic parameter names to concrete values
   */
  function expandInherits (inherits: Inherits | undefined, mappings: Map<string, ValueOf>): Inherits | undefined {
    if (inherits == null) {
      return undefined
    }
    // Similar to InstanceOf
    const expanded = expandValueOf(
      {
        kind: 'instance_of',
        type: inherits.type,
        generics: inherits.generics
      },
      mappings
    ) as InstanceOf

    return {
      type: expanded.type,
      generics: undefined
    }
  }

  /**
   * Expand a type alias
   *
   * @param alias the type definition
   * @param params concrete values for the type's generic parameters, if any
   */
  function expandTypeAlias (alias: TypeAlias, params: ValueOf[]): TypeName {
    return addIfNotSeen(expandedName(alias, params), () => {
      const result = { ...alias }
      result.type = expandValueOf(alias.type, genericParamMapping(alias.generics, params))
      result.generics = undefined
      return result
    })
  }

  /**
   * Expand a request definition
   *
   * @param req the type definition
   * @param params concrete values for the type's generic parameters, if any
   */
  function expandRequest (req: Request, params: ValueOf[]): TypeName {
    // Do not rename requests, even if they have generic params
    return addIfNotSeen(req.name, () => {
      const mappings = genericParamMapping(req.generics, params)
      const result = { ...req }
      result.inherits = expandInherits(result.inherits, mappings)
      result.path = expandProperties(result.path, mappings)
      result.query = expandProperties(result.query, mappings)
      result.body = expandBody(req.body, genericParamMapping(req.generics, params))
      result.generics = undefined
      return result
    })
  }

  /**
   * Expand a response definition
   *
   * @param resp the type definition
   * @param params concrete values for the type's generic parameters, if any
   */
  function expandResponse (resp: Response, params: ValueOf[]): TypeName {
    // Do not rename responses, even if they have generic params
    return addIfNotSeen(resp.name, () => {
      const result = { ...resp }
      result.body = expandBody(resp.body, genericParamMapping(resp.generics, params))
      if (resp.exceptions != null) {
        result.exceptions = []
        for (const exception of resp.exceptions) {
          const except: ResponseException = {
            description: exception.description,
            statusCodes: exception.statusCodes,
            body: expandBody(exception.body, genericParamMapping(resp.generics, params))
          }
          result.exceptions.push(except)
        }
      }
      result.generics = undefined
      return result
    })
  }

  /**
   * Expand an array of properties
   *
   * @param properties the properties
   * @param mappings mapping of generic parameter names to concrete values
   */
  function expandProperties (properties: Property[], mappings: Map<string, ValueOf>): Property[] {
    return properties.map(prop => {
      const newProp = { ...prop }
      newProp.type = expandValueOf(prop.type, mappings)
      return newProp
    })
  }

  /**
   * Expand a request or response body
   *
   * @param body the body definition
   * @param mappings mapping of generic parameter names to concrete values
   */
  function expandBody (body: Body, mappings: Map<string, ValueOf>): Body {
    switch (body.kind) {
      case 'no_body':
        return body

      case 'properties': {
        return {
          kind: 'properties',
          properties: expandProperties(body.properties, mappings)
        }
      }

      case 'value': {
        return {
          kind: 'value',
          value: expandValueOf(body.value, mappings),
          codegenName: body.codegenName
        }
      }
    }
  }

  /**
   * Expand a value
   *
   * @param value the value
   * @param mappings mapping of generic parameter names to concrete values
   */
  function expandValueOf (value: ValueOf, mappings: Map<string, ValueOf>): ValueOf {
    switch (value.kind) {
      case 'array_of':
        return {
          kind: 'array_of',
          value: expandValueOf(value.value, mappings)
        }

      case 'dictionary_of':
        return {
          kind: 'dictionary_of',
          key: expandValueOf(value.key, mappings),
          value: expandValueOf(value.value, mappings),
          singleKey: value.singleKey
        }

      case 'instance_of': {
        const valueOfType = nameKey(value.type)

        // If this is a type that has to be unwrapped, return its generic parameter's type
        if (typesToUnwrap.has(valueOfType)) {
          // @ts-expect-error
          const x = value.generics[0]
          return expandValueOf(x, mappings)
        }

        // If this is a type that has to be inlined
        if (typesToInline.has(valueOfType)) {
          // It has to be an alias (e.g. Stringified or WithNullValue
          const inlinedTypeDef = inputTypeByName.get(valueOfType)
          if (inlinedTypeDef?.kind !== 'type_alias') {
            throw Error(`Inlined type ${valueOfType} should be an alias definition`)
          }

          const inlineMappings = new Map<string, ValueOf>()
          for (let i = 0; i < (inlinedTypeDef.generics?.length ?? 0); i++) {
            // @ts-expect-error
            const source = inlinedTypeDef.generics[i]
            // @ts-expect-error
            const dest = value.generics[i]
            inlineMappings.set(nameKey(source), dest)
          }

          return expandValueOf(inlinedTypeDef.type, inlineMappings)
        }

        // If this is a generic parameter, return its mapping
        const mapping = mappings.get(nameKey(value.type))
        if (mapping !== undefined) {
          return mapping
        }

        // If the type is instanciated with generic parameters, expand them
        const params = (value.generics ?? []).map(g => expandValueOf(g, mappings))

        return {
          kind: 'instance_of',
          type: expandType(value.type, params),
          generics: undefined
        }
      }

      case 'literal_value':
        return value

      case 'union_of':
        return {
          kind: 'union_of',
          items: value.items.map(item => expandValueOf(item, mappings))
        }

      case 'user_defined_value':
        return value
    }
  }

  /**
   * Return a name with generic parameter names inlined
   */
  function expandedName (type: Interface | TypeAlias, params: ValueOf[]): TypeName {
    let localName = type.name.name

    type.generics?.forEach((paramName, i) => {
      const param = params[i]
      if (param.kind === 'user_defined_value') {
        return
      }
      localName = localName + valueTypeName(params[i])
    })

    return {
      namespace: type.name.namespace,
      name: localName
    }
  }

  for (const endpoint of inputModel.endpoints) {
    expandRootType(endpoint.request)
    expandRootType(endpoint.response)
  }

  // Allows to retrieve EsqlBase*EsqlVersion
  for (const type of inputModel.types) {
    addDanglingTypeIfNotSeen(type)
  }

  sortTypeDefinitions(types)

  return {
    _info: inputModel._info,
    endpoints: inputModel.endpoints,
    types: types
  }
}

/**
 * Create a mapping from generic type names to concrete values
 */
function genericParamMapping (generics: TypeName[] | undefined, params: ValueOf[]): Map<string, ValueOf> {
  const mapping = new Map<string, ValueOf>();
  (generics ?? []).forEach((name, i) => {
    mapping.set(nameKey(name), params[i])
  })

  return mapping
}

/**
 * A value's name, to be appended to the name of a type that has generic parameters
 */
function valueTypeName (value: ValueOf): string {
  switch (value.kind) {
    case 'literal_value':
      return value.value.toString()

    case 'user_defined_value':
      return 'UserDefined'

    case 'array_of':
      return 'Array' + valueTypeName(value.value)

    case 'dictionary_of':
      // Don't care about key, it's always aliased to string
      return 'Dict' + valueTypeName(value.value)

    case 'union_of':
      return 'Union' + value.items.map(v => valueTypeName(v)).join()

    case 'instance_of':
      return value.type.name
  }
}

/**
 * Return a string representing a type name, to be used as Map keys
 */
function nameKey (t: TypeDefinition | TypeName): string {
  if (isTypeDef(t)) {
    return nameKey(t.name)
  } else {
    return t.namespace + ':' + t.name
  }
}

function isTypeDef (t: any): t is TypeDefinition {
  return t.kind !== undefined
}

async function expandGenericsFromFile (inPath: string, outPath: string): Promise<void> {
  const inputText = await readFile(
    inPath,
    { encoding: 'utf8' }
  )

  const inputModel = JSON.parse(inputText)
  const outputModel = expandGenerics(inputModel, {
    // unwrappedTypes: ["_spec_utils:Stringified"],
    inlinedTypes: ['_spec_utils:WithNullValue']
  })

  await writeFile(
    outPath,
    stringify(outputModel, null, 2),
    'utf8'
  )
}

const inputPath = argv.input ?? join(__dirname, '..', '..', '..', 'output', 'schema', 'schema.json')
const outputPath = argv.output ?? join(__dirname, '..', '..', '..', 'output', 'schema', 'schema-no-generics.json')

expandGenericsFromFile(inputPath, outputPath)
  .catch(reason => console.error(reason))
  .finally(() => console.log('Done, expanded schema is at', outputPath))

File: compiler-rs/clients_schema/src/transform/expand_generics.rs
// Licensed to Elasticsearch B.V. under one or more contributor
// license agreements. See the NOTICE file distributed with
// this work for additional information regarding copyright
// ownership. Elasticsearch B.V. licenses this file to you under
// the Apache License, Version 2.0 (the "License"); you may
// not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

use std::collections::{HashMap, HashSet};

use anyhow::bail;
use indexmap::IndexMap;

use crate::*;

#[derive(Default)]
struct Ctx {
    new_types: IndexMap<TypeName, TypeDefinition>,
    types_seen: std::collections::HashSet<TypeName>,
    config: ExpandConfig,
}

/// Generic parameters of a type
type GenericParams = Vec<TypeName>;
/// Generic arguments for an instantiated generic type
type GenericArgs = Vec<ValueOf>;
/// Mapping from generic arguments to values
type GenericMapping = HashMap<TypeName, ValueOf>;

#[derive(Clone, Debug)]
pub struct ExpandConfig {
    /// Generic types that will be inlined by replacing them with their definition, propagating generic arguments.
    pub unwrap: HashSet<TypeName>,
    // Generic types that will be unwrapped by replacing them with their (single) generic parameter.
    pub inline: HashSet<TypeName>,
}

impl Default for ExpandConfig {
    fn default() -> Self {
        ExpandConfig {
            unwrap: Default::default(),
            inline: HashSet::from([builtins::WITH_NULL_VALUE])
        }
    }
}

/// Expand all generics by creating new concrete types for every instantiation of a generic type.
///
/// The resulting model has no generics anymore. Top-level generic parameters (e.g. SearchRequest's TDocument) are
/// replaced by user_defined_data.
pub fn expand(model: IndexedModel, config: ExpandConfig) -> anyhow::Result<IndexedModel> {
    let mut model = model;
    let mut ctx = Ctx {
        config,
        ..Ctx::default()
    };

    for endpoint in &model.endpoints {
        for name in [&endpoint.request, &endpoint.response].into_iter().flatten() {
            expand_root_type(name, &model, &mut ctx)?;
        }
    }

    // Add new types that were created to the model
    ctx.new_types.sort_keys();
    model.types = ctx.new_types;

    return Ok(model);

    //---------------------------------------------------------------------------------------------
    // Expanding types
    //---------------------------------------------------------------------------------------------

    fn expand_root_type(t: &TypeName, model: &IndexedModel, ctx: &mut Ctx) -> anyhow::Result<()> {
        const NO_GENERICS: &Vec<TypeName> = &Vec::new();
        const USER_DEFINED: ValueOf = ValueOf::UserDefinedValue(UserDefinedValue {});

        use TypeDefinition::*;
        let generics = match model.get_type(t)? {
            Interface(itf) => &itf.generics,
            Request(req) => &req.generics,
            Response(resp) => &resp.generics,
            Enum(_) | TypeAlias(_) => NO_GENERICS,
        };

        // Top-level generic parameters (e.g. TDocument in SearchResponse) are set to UserDefined
        let args: GenericArgs = generics.iter().map(|_| USER_DEFINED).collect();

        expand_type(t, args, model, ctx)?;

        Ok(())
    }

    /// Expand a type definition, given concrete values for its generic parameters.
    /// The new type definition is stored in the context.
    ///
    /// Returns the name to use for this (type, args) combination
    fn expand_type(
        name: &TypeName,
        args: GenericArgs,
        model: &IndexedModel,
        ctx: &mut Ctx,
    ) -> anyhow::Result<TypeName> {
        if name.is_builtin() {
            return Ok(name.clone());
        }

        let def = model.get_type(name)?;
        let name = expanded_name(def.name(), &args);

        if !ctx.types_seen.contains(&name) {
            // Mark it as seen to avoid infinite recursion
            ctx.types_seen.insert(name.clone());

            let mut new_def = match def {
                TypeDefinition::Interface(ref itf) => expand_interface(itf, args, model, ctx)?,
                TypeDefinition::Request(req) => expand_request(req, args, model, ctx)?,
                TypeDefinition::Response(resp) => expand_response(resp, args, model, ctx)?,
                TypeDefinition::TypeAlias(ref alias) => expand_type_alias(alias, args, model, ctx)?,
                TypeDefinition::Enum(_) => def.clone(),
            };
            new_def.base_mut().name = name.clone();
            ctx.new_types.insert(name.clone(), new_def);
        }

        Ok(name)
    }

    fn expand_interface(
        itf: &Interface,
        args: GenericArgs,
        model: &IndexedModel,
        ctx: &mut Ctx,
    ) -> anyhow::Result<TypeDefinition> {
        // Clone and modify in place
        let mut itf = itf.clone();

        let mappings = param_mapping(&itf.generics, args);

        itf.generics = Vec::new();

        if let Some(inherit) = itf.inherits {
            itf.inherits = Some(expand_inherits(inherit, &mappings, model, ctx)?);
        }

        if !itf.behaviors.is_empty() {
            itf.behaviors.iter().for_each(|b| {
                if b.generics.is_empty() {
                    // If the behavior has no generics, we can just expand it
                    let _ = expand_type(&b.typ, Vec::new(), model, ctx);
                }
            });
        }

        // We keep the generic parameters of behaviors, but expand their value
        for behavior in &mut itf.behaviors {
            for arg in &mut behavior.generics {
                *arg = expand_valueof(arg, &mappings, model, ctx)?;
            }
        }

        expand_properties(&mut itf.properties, &mappings, model, ctx)?;

        Ok(itf.into())
    }

    fn expand_request(
        req: &Request,
        args: GenericArgs,
        model: &IndexedModel,
        ctx: &mut Ctx,
    ) -> anyhow::Result<TypeDefinition> {
        // Clone and modify in place
        let mut req = req.clone();

        let mappings = param_mapping(&req.generics, args);
        req.generics = Vec::new();

        if let Some(inherit) = req.inherits {
            req.inherits = Some(expand_inherits(inherit, &mappings, model, ctx)?);
        }

        expand_behaviors(&mut req.behaviors, &mappings, model, ctx)?;
        expand_properties(&mut req.path, &mappings, model, ctx)?;
        expand_properties(&mut req.query, &mappings, model, ctx)?;
        expand_body(&mut req.body, &mappings, model, ctx)?;

        Ok(req.into())
    }

    fn expand_response(
        resp: &Response,
        args: GenericArgs,
        model: &IndexedModel,
        ctx: &mut Ctx,
    ) -> anyhow::Result<TypeDefinition> {
        // Clone and modify in place
        let mut resp = resp.clone();

        let mappings = param_mapping(&resp.generics, args);
        resp.generics = Vec::new();

        expand_behaviors(&mut resp.behaviors, &mappings, model, ctx)?;
        expand_body(&mut resp.body, &mappings, model, ctx)?;
        
        for exception in &mut resp.exceptions {
            expand_body(&mut exception.body, &mappings, model, ctx)?;
        }

        Ok(resp.into())
    }

    fn expand_type_alias(
        t: &TypeAlias,
        args: GenericArgs,
        model: &IndexedModel,
        ctx: &mut Ctx,
    ) -> anyhow::Result<TypeDefinition> {
        let mapping = param_mapping(&t.generics, args);
        let value = expand_valueof(&t.typ, &mapping, model, ctx)?;

        Ok(TypeDefinition::TypeAlias(TypeAlias {
            base: t.base.clone(),
            generics: Vec::new(),
            typ: value,
            variants: t.variants.clone(),
        }))
    }

    //---------------------------------------------------------------------------------------------
    // Expanding type parts in place
    //---------------------------------------------------------------------------------------------

    fn expand_inherits(
        i: Inherits,
        mappings: &GenericMapping,
        model: &IndexedModel,
        ctx: &mut Ctx,
    ) -> anyhow::Result<Inherits> {
        let expanded = expand_valueof(
            &InstanceOf {
                typ: i.typ,
                generics: i.generics,
            }
            .into(),
            mappings,
            model,
            ctx,
        )?;

        if let ValueOf::InstanceOf(inst) = expanded {
            Ok(Inherits {
                typ: inst.typ,
                generics: Vec::new(),
            })
        } else {
            bail!("Inherits clause doesn't expand to an instance_of: {:?}", &expanded);
        }
    }

    /// Expand behaviors in place
    fn expand_behaviors(
        behaviors: &mut Vec<Inherits>,
        mappings: &GenericMapping,
        model: &IndexedModel,
        ctx: &mut Ctx,
    ) -> anyhow::Result<()> {
        // We keep the generic parameters of behaviors, but expand their value
        for behavior in behaviors {
            for arg in &mut behavior.generics {
                *arg = expand_valueof(arg, mappings, model, ctx)?;
            }
        }
        Ok(())
    }

    /// Expand properties in place
    fn expand_properties(
        props: &mut Vec<Property>,
        mappings: &GenericMapping,
        model: &IndexedModel,
        ctx: &mut Ctx,
    ) -> anyhow::Result<()> {
        for prop in props {
            prop.typ = expand_valueof(&prop.typ, mappings, model, ctx)?;
        }
        Ok(())
    }

    /// Expand body in place
    fn expand_body(
        body: &mut Body,
        mappings: &GenericMapping,
        model: &IndexedModel,
        ctx: &mut Ctx,
    ) -> anyhow::Result<()> {
        match body {
            Body::Value(ref mut value) => {
                value.value = expand_valueof(&value.value, mappings, model, ctx)?;
            }
            Body::Properties(ref mut prop_body) => {
                expand_properties(&mut prop_body.properties, mappings, model, ctx)?;
            }
            Body::NoBody(_) => {}
        }

        Ok(())
    }

    //---------------------------------------------------------------------------------------------
    // Expanding values
    //---------------------------------------------------------------------------------------------

    fn expand_valueof(
        value: &ValueOf,
        mappings: &GenericMapping,
        model: &IndexedModel,
        ctx: &mut Ctx,
    ) -> anyhow::Result<ValueOf> {
        match value {
            ValueOf::ArrayOf(ref arr) => {
                let value = expand_valueof(&arr.value, mappings, model, ctx)?;
                Ok(ArrayOf { value: Box::new(value) }.into())
            }

            ValueOf::DictionaryOf(dict) => {
                let key = expand_valueof(&dict.key, mappings, model, ctx)?;
                let value = expand_valueof(&dict.value, mappings, model, ctx)?;
                Ok(DictionaryOf {
                    single_key: dict.single_key,
                    key: Box::new(key),
                    value: Box::new(value),
                }
                .into())
            }

            ValueOf::InstanceOf(inst) => {
                // If this is a generic parameter, return its mapping
                if let Some(p) = mappings.get(&inst.typ) {
                    return Ok(p.clone());
                }

                // Inline or unwrap if required by the config
                if ctx.config.inline.contains(&inst.typ) {
                    return inline_generic_type(inst, mappings, model, ctx);
                }
                if ctx.config.unwrap.contains(&inst.typ) {
                    return unwrap_generic_type(inst, mappings, model, ctx);
                }

                // Expand generic parameters, if any
                let args = inst
                    .generics
                    .iter()
                    .map(|arg| expand_valueof(arg, mappings, model, ctx))
                    .collect::<Result<Vec<_>, _>>()?;

                Ok(InstanceOf {
                    typ: expand_type(&inst.typ, args, model, ctx)?,
                    generics: Vec::new(),
                }
                .into())
            }

            ValueOf::UnionOf(u) => {
                let items = u
                    .items
                    .iter()
                    .map(|item| expand_valueof(item, mappings, model, ctx))
                    .collect::<Result<Vec<_>, _>>()?;
                Ok(UnionOf { items }.into())
            }

            ValueOf::UserDefinedValue(_) => Ok(value.clone()),

            ValueOf::LiteralValue(_) => Ok(value.clone()),
        }
    }

    /// Inlines a value of a generic type by replacing it with its definition, propagating
    /// generic arguments.
    fn inline_generic_type(
        value: &InstanceOf,
        _mappings: &GenericMapping,
        model: &IndexedModel,
        ctx: &mut Ctx,
    ) -> anyhow::Result<ValueOf> {

        // It has to be an alias (e.g. WithNullValue)
        if let TypeDefinition::TypeAlias(inline_def) = model.get_type(&value.typ)? {
            // Create mappings to resolve types in the inlined type's definition
            let mut inline_mappings = GenericMapping::new();
            for (source, dest) in inline_def.generics.iter().zip(value.generics.iter()) {
                inline_mappings.insert(source.clone(), dest.clone());
            }
            // and expand the inlined type's alias definition
            let result = expand_valueof(&inline_def.typ, &inline_mappings, model, ctx)?;
            return Ok(result);
        } else {
            bail!("Expecting inlined type {} to be an alias", &value.typ);
        }
    }

    /// Unwraps a value of a generic type by replacing it with its generic parameter
    fn unwrap_generic_type(
        value: &InstanceOf,
        mappings: &GenericMapping,
        model: &IndexedModel,
        ctx: &mut Ctx,
    ) -> anyhow::Result<ValueOf> {

        // It has to be an alias (e.g. Stringified)
        if let TypeDefinition::TypeAlias(_unwrap_def) = model.get_type(&value.typ)? {
            // Expand the inlined type's generic argument (there must be exactly one)
            if value.generics.len() != 1 {
                bail!("Expecting unwrapped type {} to have exactly one generic parameter", &value.typ);
            }
            let result = expand_valueof(&value.generics[0], mappings, model, ctx)?;
            return Ok(result);
        } else {
            bail!("Expecting unwrapped type {} to be an alias", &value.typ);
        }
    }

    //---------------------------------------------------------------------------------------------
    // Misc
    //---------------------------------------------------------------------------------------------

    /// Builds the mapping from generic parameter name to actual value
    fn param_mapping(generics: &GenericParams, args: GenericArgs) -> GenericMapping {
        generics.iter().cloned().zip(args).collect()
    }

    /// Creates an expanded type name if needed (i.e. when `generics` is not empty)
    fn expanded_name(type_name: &TypeName, args: &GenericArgs) -> TypeName {
        if args.is_empty() {
            return type_name.clone();
        }

        let mut name: String = type_name.name.to_string();
        for arg in args {
            if let ValueOf::UserDefinedValue(_) = arg {
                // Top-level types. Don't append it.
            } else {
                push_valueof_name(&mut name, arg);
            }
        }

        TypeName {
            namespace: type_name.namespace.clone(),
            name: name.into(),
        }
    }

    /// Appends the type representation of a value to a string
    fn push_valueof_name(name: &mut String, value: &ValueOf) {
        use std::fmt::Write;
        match value {
            ValueOf::LiteralValue(lit) => write!(name, "{}", lit).unwrap(),
            ValueOf::UserDefinedValue(_) => write!(name, "UserDefined").unwrap(),
            ValueOf::ArrayOf(a) => {
                name.push_str("Array");
                push_valueof_name(name, a.value.as_ref());
            }
            ValueOf::DictionaryOf(dict) => {
                // Don't care about key, it's always aliased to string
                name.push_str("Dict");
                push_valueof_name(name, dict.value.as_ref())
            }
            ValueOf::UnionOf(u) => {
                name.push_str("Union");
                for item in &u.items {
                    push_valueof_name(name, item)
                }
            }
            ValueOf::InstanceOf(inst) => {
                // Append unqualified name (assuming we have no duplicate generic value for the same type)
                name.push_str(&inst.typ.name);
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use std::io::Write;

    use super::*;

    #[test]
    pub fn compare_with_js_version() -> testresult::TestResult {
        let canonical_json = {
            // Deserialize and reserialize to have a consistent JSON format
            let json = std::fs::read_to_string("../../output/schema/schema-no-generics.json")?;
            let model: IndexedModel = serde_json::from_str(&json)?;
            serde_json::to_string_pretty(&model)?
        };

        let schema_json = std::fs::read_to_string("../../output/schema/schema.json")?;
        let model: IndexedModel = serde_json::from_str(&schema_json)?;
        let model = expand(model, ExpandConfig::default())?;

        let json_no_generics = serde_json::to_string_pretty(&model)?;

        if canonical_json != json_no_generics {
            std::fs::create_dir_all("test-output")?;
            let mut out = std::fs::File::create("test-output/schema-no-generics-canonical.json")?;
            out.write_all(canonical_json.as_bytes())?;

            let mut out = std::fs::File::create("test-output/schema-no-generics-new.json")?;
            out.write_all(json_no_generics.as_bytes())?;

            panic!("Output differs from the canonical version. Both were written to 'test-output'");
        }

        Ok(())
    }
}
